reference-gen/pkg/generator/generator.go

```diff
@@ -0,0 +1,187 @@
+package generator
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"text/template"
+
+	"k8s.io/gengo/types"
+	"k8s.io/klog/v2"
+)
+
+const (
+	generatedTextWarning = "<!--- THIS FILE IS AUTOGENERATED!!! DO NOT EDIT!!! -->\n"
+)
+
+type Generator interface {
+	Run() error
+}
+
+func NewGenerator(packageName string, requestedTypesList []string, headerTextFile string, outputFileName string, templateDirectory string) (Generator, error) {
+	if packageName == "" {
+		return nil, errors.New("a package name must be specified")
+	}
+
+	headerText, err := loadHeaderText(headerTextFile)
+	if err != nil {
+		return nil, fmt.Errorf("error loading header text: %v", err)
+	}
+
+	if err := checkTemplateDir(templateDirectory); err != nil {
+		return nil, fmt.Errorf("invalid template directory: %v", err)
+	}
+
+	return &generator{
+		packageName:       packageName,
+		requestedTypes:    newStringSet(requestedTypesList),
+		headerText:        headerText,
+		outputFileName:    outputFileName,
+		templateDirectory: templateDirectory,
+	}, nil
+}
+
+// checkTemplateDir checks whether the template directory given exists and can be read
+func checkTemplateDir(dir string) error {
+	if dir == "" {
+		return nil
+	}
+	path, err := filepath.Abs(dir)
+	if err != nil {
+		return err
+	}
+
+	if fi, err := os.Stat(path); err != nil {
+		return fmt.Errorf("cannot read directory %q: %v", path, err)
+	} else if !fi.IsDir() {
+		return errors.New("path %q is not a directory")
```

Comment:
Did you mean for `fmt.Errorf` with `path` for that `%q`?

Reply:
100%, yes I did, oops 😂